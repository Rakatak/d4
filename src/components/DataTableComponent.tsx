import React, { ReactElement, useState } from "react";
import DataService from '../services/DataService';
import DataTable from 'react-data-table-component';
import { FilterComponent } from './FilterComponent';

// interface with typings for GovData respone 
interface IGovData {
    department: string,
    description: string,
    datasets: number
}

// configuring columns for DataTable, enabling sorting      
const columns = [
    {
        name: 'Department',
        selector: (row: { department: string; }) => row.department,
        sortable: true,
    },
    {
        name: 'Description',
        selector: (row: { description: string; }) => row.description,
        sortable: true,
    },
    {
        name: 'Description',
        selector: (row: { datasets: number; }) => row.datasets,
        sortable: true,
    },
];


export const DataTableComponent = (): ReactElement => {
    const [govData, setGovData] = useState<IGovData[]>([]);
    const [filterText, setFilterText] = React.useState('');

    DataService.getInstance().getGovData().then((result: IGovData[]) => {
        console.log('result from API ====>', result)
        setGovData(result);
    });

    // get all departments after filtering with textfield input data
    const filteredItems = govData.filter(
        item => item.department && item.department.toLowerCase().includes(filterText.toLowerCase()),
    );

    const subHeaderComponentMemo = React.useMemo(() => {
        const handleClear = () => {
            if (filterText) {
                setFilterText('');
            }
        };

        return (
            <FilterComponent onFilter={(e: { target: { value: React.SetStateAction<string>; }; }) => setFilterText(e.target.value)} onClear={handleClear} filterText={filterText} />
        );
    }, [filterText]);

    return (
        <DataTable
            title="GovData Table"
            columns={columns}
            data={filteredItems}
            subHeader
            subHeaderComponent={subHeaderComponentMemo}
            persistTableHead
        />
    );
};