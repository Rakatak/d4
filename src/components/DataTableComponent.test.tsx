import { DataTableComponent } from './DataTableComponent';
import renderer from 'react-test-renderer';

describe('Test for DataTable', () => {
  it('Renders the data table component and tests the snapshot', () => {
    const component = renderer.create(
      <DataTableComponent/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

