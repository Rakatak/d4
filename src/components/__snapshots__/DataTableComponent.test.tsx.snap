// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for DataTable Renders the data table component and tests the snapshot 1`] = `
Array [
  <div
    aria-level={1}
    className="sc-fEOsli cEMZEX rdt_TableHeader"
    role="heading"
  >
    <div
      className="sc-bjUoiL kWAsOx"
    >
      GovData Table
    </div>
    <div
      className="sc-fnykZs kIZdof"
    >
      <div
        className="sc-ksZaOG eAWRvo"
      />
      <div
        className="sc-hAZoDl jEpWRK"
      />
    </div>
  </div>,
  <header
    className="sc-dIouRR TUEFO"
  >
    <input
      aria-label="Search Input"
      className="sc-jIZahH jYrXcl"
      id="search"
      onChange={[Function]}
      placeholder="Filter By Name"
      type="text"
      value=""
    />
    <button
      className="sc-himrzO kkoyas"
      onClick={[Function]}
      type="button"
    >
      X
    </button>
  </header>,
  <div
    className="sc-dmRaPn gelpCx"
  >
    <div
      className="sc-fLlhyt ifOHjV"
    >
      <div
        className="sc-bczRLJ eSTlnH rdt_Table"
        disabled={false}
        role="table"
      >
        <div
          className="sc-gsnTZi idPHNo rdt_TableHead"
          role="rowgroup"
        >
          <div
            className="sc-dkzDqf gZWIKX rdt_TableHeadRow"
            role="row"
          >
            <div
              className="sc-hKMtZM sc-eCYdqJ sc-iqcoie jzGaAN bnAwAJ crSOFH rdt_TableCol"
              data-column-id={1}
              onDragEnd={[Function]}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDragStart={[Function]}
            >
              <div
                className="sc-crXcEl bLyqNT rdt_TableCol_Sortable"
                data-column-id={1}
                data-sort-id={1}
                disabled={true}
                role="columnheader"
                tabIndex={0}
              >
                <div
                  className="sc-evZas eUeqdG"
                  data-column-id={1}
                >
                  Department
                </div>
              </div>
            </div>
            <div
              className="sc-hKMtZM sc-eCYdqJ sc-iqcoie jzGaAN bnAwAJ crSOFH rdt_TableCol"
              data-column-id={2}
              onDragEnd={[Function]}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDragStart={[Function]}
            >
              <div
                className="sc-crXcEl bLyqNT rdt_TableCol_Sortable"
                data-column-id={2}
                data-sort-id={2}
                disabled={true}
                role="columnheader"
                tabIndex={0}
              >
                <div
                  className="sc-evZas eUeqdG"
                  data-column-id={2}
                >
                  Description
                </div>
              </div>
            </div>
            <div
              className="sc-hKMtZM sc-eCYdqJ sc-iqcoie jzGaAN bnAwAJ crSOFH rdt_TableCol"
              data-column-id={3}
              onDragEnd={[Function]}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDragStart={[Function]}
            >
              <div
                className="sc-crXcEl bLyqNT rdt_TableCol_Sortable"
                data-column-id={3}
                data-sort-id={3}
                disabled={true}
                role="columnheader"
                tabIndex={0}
              >
                <div
                  className="sc-evZas eUeqdG"
                  data-column-id={3}
                >
                  Description
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="sc-ivTmOn fwKvpK"
        >
          <div
            style={
              Object {
                "padding": "24px",
              }
            }
          >
            There are no records to display
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
